### DISCUSSION TITLE
Download js/css from cdn only when progressively hydrating a component (client:visible)

### DISCUSSION THREAD

**Discussion by TechAkayy** [ElMassimo/iles#267] [General]
Hello islanders!

On a simple html page, using [11ty/is-land](https://github.com/11ty/is-land) package, I was able to trigger the download of the js & css from the cdn as part of the progressive hydration of the island with the `on:visible` directive of this package (same as our `client:visible`). Here is the html page - https://github.com/TechAkayy/prozone-chairs-html (demo - https://prozone-chairs-html.netlify.app/)

Is this achievable with Iles?

Please refer to this sample repo, that's pretty much the same html page but components needing hydration are pulled out as separate components - https://github.com/TechAkayy/prozone-chairs-iles (current demo - https://prozone-chairs-iles.netlify.app/)

I'm progressively hydrating my `BookAnAppointment` component with `client:visible` in my index page. This component uses [pikaday](https://github.com/Pikaday/Pikaday) datepicker. 

![image](https://github.com/ElMassimo/iles/assets/73772701/d0f1dc04-b7b4-48d6-8c57-2b46cfd5db2e)

Pikaday js & css can be used via npm import, though I want to use it from the cdn, as a way of learning how to apply this technique to libraries that are only available via cdn. I'm currently adding the cdn links via `app.ts` by [extending](https://iles-docs.netlify.app/config#your-app) the head tag. This works just fine, the component hydrates when entering the viewport. 

<img width="462" alt="image" src="https://github.com/ElMassimo/iles/assets/73772701/798cce33-5507-4380-af1f-f89cf794b102">

But, my question is if there is a way to only download the pikaday js & css from the cdn only when the component is hydrating via client:visible directive. Let me know if it's possible to achieve  this.

Some attempts from my end: 
- [`<Head>`](https://iles-docs.netlify.app/guide/meta-tags) component wasn't working as Vite ignores the script tags added as side-effects. 

- I tried adding them to the head tag using [Client Scripts](https://iles-docs.netlify.app/guide/client-scripts) SFC custom section, but this doesn't work nicely with the `client:visible` on the component, and I'm getting the `appConfig.js?v=78c168bf:9 Uncaught (in promise) Error: App config not properly injected in app` error. 
<img width="601" alt="image" src="https://github.com/ElMassimo/iles/assets/73772701/6057f58a-9aee-4f42-9573-c8420c3e0630">

- If I remove the `client:visible` on it, I then get the Pikaday is not defined error. Probably more of a usage issue - as to how to couple using cdn imports via client script in the component.
<img width="1192" alt="image" src="https://github.com/ElMassimo/iles/assets/73772701/7d8a73c5-3c58-4cb3-8366-e2683b394b30">

**Comment by ElMassimo** [comment]
Hi!

[`useScriptTag`](https://vueuse.org/core/useScriptTag/) would allow you to load scripts outside of the bundle. By calling it in the component you would ensure that the script is only loaded when the component is created (on hydration).

**Comment by TechAkayy** [comment]
Perfect, just what I wanted. On a related topic, is there an easy way to add cdn links to the head tag without extending the head via `app.ts`? Something similar to what Astro does, where we can paste the script tag in the layout file, and mark it with `is:inline`.

**Reply by ElMassimo** [reply]
Unfortunately, Vue doesn't like `script` tags in the template, it will raise a warning if you try that.

Passing `script` to [`useHead`](https://github.com/ElMassimo/iles/blob/vite-5/playground/the-vue-point/src/layouts/default.vue#L4) (or [`useScript`](https://unhead.unjs.io/usage/composables/use-script) once I finish the release using `unhead`) will work if you use it in your layouts or non-interactive components.

For usage inside islands, [`useScriptTag`](https://vueuse.org/core/useScriptTag/) is the best option.

**Reply by TechAkayy** [reply]
Thanks @ElMassimo!

**Reply by TechAkayy** [reply]
Hi @ElMassimo, 
Is it possible to use an index.html in the root folder that can act as the entry file? This way adding cdn links will be easier similar to in html.

**Reply by ElMassimo** [reply]
Not at the moment, but it's planned:

- #136

**Reply by TechAkayy** [reply]
That's awesome, this could be a easy way to add external resources, as it's a familiar file in vanilla vue setup. This will help new-to-iles users big time.
